Pinta.Effects/Effects/ForwardErrorDiffusionDitheringEffect.cs

```diff
@@ -0,0 +1,325 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Cairo;
+using Pinta.Core;
+using Pinta.Gui.Widgets;
+
+namespace Pinta.Effects;
+
+public sealed class ForwardErrorDiffusionDitheringEffect : BaseEffect
+{
+	public override string Name => Translations.GetString ("Dithering");
+	public override bool IsConfigurable => true;
+	// TODO: Icon
+	public override string EffectMenuCategory => Translations.GetString ("Test"); // TODO:
+	public ForwardErrorDiffusionDitheringData Data => (ForwardErrorDiffusionDitheringData) EffectData!; // NRT - Set in constructor
+
+	public ForwardErrorDiffusionDitheringEffect ()
+	{
+		EffectData = new ForwardErrorDiffusionDitheringData ();
+	}
+
+	public override void Render (ImageSurface src, ImageSurface dest, ReadOnlySpan<RectangleI> rois)
+	{
+		var src_data = src.GetReadOnlyPixelData ();
+		var src_copy = new ColorBgra[src_data.Length];
```

Comment:
The article you linked to mentioned that you only need a couple rows to store the errors accumulated for the current + next row (varies based the matrix of course), so that could be a future optimization.

Or, at least avoid the extra copy by copying into the destination image and then working in-place there